WIP: OCPBUGS-61756: ctrl: kubeletConfig: avoid updating ConfigMap on Paused MCP

shajmakh opened the pull request:
Having a paused MCP should prevent updating/creating the corresponding
config map for the specified node group. So far, the code wasn't
considering the case of paused MCPs, which leads to creating/updating the config map
a thing that caused a mismatch between the configuration in the config
map and the one reflected on the NRTs.
In this commit, we modify the kubeletconfig controller to skip on
updating RTE config maps that belong to paused MCPs.

Tal-or commented on internal/reconcile/step.go:
```diff
@@ -92,3 +101,35 @@ func StepFailed(err error) Step {
 	}
 }
+
+func StepWarning(err error) Step {
```
lets add comments explaining the new steps purposes.

shajmakh replied:
of course

Tal-or reviewed (commented):
Good change overall.

In case the MCP is paused, the final update to the configmap will be triggered when the MCP changes back to unpause?

shajmakh commented on internal/reconcile/event.go:
```diff
@@ -0,0 +1 @@
+package reconcile
```
correct, was thinking if we need a different wrapper other than the one we use for NRO controller, for now we can use the same `step` record

shajmakh commented:
/retest

shajmakh committed:
ctrl: nro: handle paused MCPs

The situtaion is mainly mitigated in the kubeletconfig controller,
however we have a bug in the NRO controller such that it puts the NRO CR
in progressing state because paused MCP is not in updated condition.
This commit ignores checking whether a paused MCP is up-to-date, and
introduces a new status condition to report paused MCPs if exist.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

yanirq commented:
/retest

yanirq commented:
also fixes https://issues.redhat.com/browse/OCPBUGS-52859

ffromani reviewed (commented):
initial review

Tal-or commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -244,63 +249,106 @@ func (r *KubeletConfigReconciler) syncConfigMap(ctx context.Context, kubeletConf
+		if mcp.Spec.Paused {
+			klog.InfoS("detected paused MCP", "name", mcp.Name)
+			//if the CM exists -> just skip;
+			//if the CM does not exist -> create it based on the current active machineConfig
```
Maybe I'm not following but how such case is possible? is it the default `01-worker-kubelet` machineConfig?

shajmakh commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -244,63 +249,106 @@ func (r *KubeletConfigReconciler) syncConfigMap(ctx context.Context, kubeletConf
+					}
+
+					successStepWithRetry := intreconcile.StepNormalSucess(fmt.Sprintf("Created ConfigMap based on the current machineConfig for paused MCP %s", mcp.Name))
+					successStepWithRetry.Result = ctrl.Result{Requeue: true, RequeueAfter: MachineConfigPoolPausedRetryPeriod}
```
that sounds better, Thanks for highliting, will update this part

shajmakh committed:
ctrl: kubeletconfig: introduce reconcileErrorHandler

This is a preliminary step to expand the scenarios where the reconciliation can return more possible errors where some are tolerable.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh committed:
ctrl: kubeletConfig: handle paused MCPs

Having a paused MCP should prevent updating the corresponding
config map for the specified node group. So far, the code wasn't
considering the case of paused MCPs, which lead to creating/updating the config map
to the newest kubeletconfig CR updates,a thing that caused a mismatch
 between the configuration in the config map vs the one reflected on the NRTs.
In this commit, we modify the kubeletconfig controller to handle paused
MCPs such that it skips updating existing RTE config maps; and for new node
groups whose MCP is paused, the controller will fetch the old
machineConfig (before the pause) and creates RTE config map based on the
decoded kubeletconfig data from it.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh committed:
ctrl:kubeletconfig: use MCP predicate instead of requeue

We want to know when a paused MCP goes unpaused and do the needed
updates to the config map when that happens.
In previous commit, when a paused MCP was detected, the reconcile would
requeue after some time; that was intended to catch the unpause
operation eventually. In this commit, we don't requeue on detected
paused MCP but rather watch for pause field updates on targeted MCPs.

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh committed:
ctrl: nro: reconcile on node group with poolName

Node group can be set with either MCPselector or a poolName. Validation
of the nodegroup configuraion is checked in the reconcile logic, and
because poolName is set and MCPSelector is not a valid nodeGroup we need
to consider this and reconcile.

This was missed when nodeGroup.PoolName was first introduced and didn't
cause any observed reconcile failures because:
1. any change on the nodeGroups would anyway trigger reconciliation
2. so far the MCP related updates were caused by Kubeletconfig updates
   which wasn't blocked until the paused-MCP case got handled (which is
why this appeared now).

Signed-off-by: Shereen Haj <shajmakh@redhat.com>

shajmakh commented:
/retest

ffromani commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -87,6 +90,12 @@ type kubeletConfigHandler struct {
 	setCtrlRef func(owner, controlled metav1.Object, scheme *runtime.Scheme, opts ...controllerutil.OwnerReferenceOption) error
 }
 
+type reconcileErrorHandler struct {
```
this is not a handler. a Handler manages something. This seems more a `reconcileResult`

Tal-or commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -126,21 +129,22 @@ func (r *KubeletConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	r.Recorder.Event(instance, "Normal", "ProcessOK", fmt.Sprintf("Updated RTE config %s/%s from kubelet config %s", cm.Namespace, cm.Name, req.NamespacedName.String()))
-	return ctrl.Result{}, nil
+	//return ctrl.Result{}, nil
```
leftovers I presume

ffromani commented on internal/controller/kubeletconfig_controller.go:
```diff
@@ -254,63 +261,128 @@ func (r *KubeletConfigReconciler) syncConfigMap(ctx context.Context, kubeletConf
+				}, reconcileErrorHandler{result: ctrl.Result{Requeue: true, RequeueAfter: MachineConfigPoolPausedRetryPeriod}}
+			}
+
+			klog.InfoS("MachineConfigPool of KubeletConfig %s is paused and configMap %s exists", kcKey.Name, existingCM.Name)
```
please check verbosity to avoid log spam

ffromani commented on internal/controller/kubeletconfig_controller_test.go:
```diff
@@ -272,4 +272,171 @@ var _ = Describe("Test KubeletConfig Reconcile", func() {
+				Namespace: testNamespace,
+				Name:      objectnames.GetComponentName(nro.Name, poolName2),
+			}
+			Expect(reconciler.Client.Get(context.TODO(), key, cm)).To(HaveOccurred())
```
and check IsNotFound perhaps?

ffromani reviewed (commented):
the direction seems fine, I have implementation comments.
The last commit would need another review round from me, need more time to grasp it.

Tal-or reviewed (commented):
Some comments inside.
One thing that bother me the most is the `tolerateError` concept.
if it's tolerable, I wouldn't consider it an error.

can't we simply log the error in place, and Requeue the request as we usually do?
